src/nbl/asset/interchange/CIESProfileLoader.cpp

```diff
@@ -0,0 +1,226 @@
+
+using namespace nbl;
+using namespace asset;
+
+
+double CIESProfile::sample(double vAngle, double hAngle) const {
+    vAngle = fmod(vAngle, vAngles.back()); // warp angle
+    hAngle = hAngles.back() == 0.0
+                 ? 0.0
+                 : fmod(hAngle,
+                        hAngles.back()); // when last horizontal angle is zero
+                                         // it's symmetric across all planes
+
+    // bilinear interpolation
+    auto lb = [](const core::vector<double> &angles, double angle) -> int {
+      int idx = std::upper_bound(std::begin(angles), std::end(angles), angle) -
+                std::begin(angles);
+      return std::max(idx-1, 0);
+    };
+    auto ub = [](const core::vector<double> &angles, double angle) -> int {
+      int idx = std::upper_bound(std::begin(angles), std::end(angles), angle) -
+                std::begin(angles);
+      return std::min(idx, (int)angles.size() - 1);
+    };
+    int j0 = lb(vAngles, vAngle);
+    int j1 = ub(vAngles, vAngle);
+    int i0 = lb(hAngles, hAngle);
+    int i1 = ub(hAngles, hAngle);
+    double uResp = i1 == i0 ? 1.0 : 1.0 / (hAngles[i1] - hAngles[i0]);
+    double vResp = j1 == j0 ? 1.0 : 1.0 / (vAngles[j1] - vAngles[j0]);
+    double u = (hAngle - hAngles[i0]) * uResp;
+    double v = (vAngle - vAngles[j0]) * vResp;
+    double s0 = getValue(i0, j0) * (1.0-v) + getValue(i0, j1) * (v);
+    double s1 = getValue(i1, j0) * (1.0-v) + getValue(i1, j1) * (v);
+    return s0 * (1.0 - u) + s1 * u;
+}
+
+
+ bool CIESProfileParser::parse(CIESProfile& result) {
+     // skip metadata
+     std::string line;
+
+     while (std::getline(ss, line)) {
+         if (line.back() == '\r')
+             line.pop_back();
+         if (line == "TILT=INCLUDE" || line == "TILT=NONE")
+             break;
+     }
+
+     if (line == "TILT=INCLUDE") {
+         double lampToLuminaire = getDouble("lampToLuminaire truncated");
+         int numTilt = getDouble("numTilt truncated");
+         for (int i = 0; i < numTilt; i++)
+             getDouble("tilt angle truncated");
+         for (int i = 0; i < numTilt; i++)
+             getDouble("tilt multiplying factor truncated");
+     }
+     else if (line != "TILT=NONE") {
+         errorMsg = "TILT not specified";
+         return false;
+     }
+
+     int numLamps = getInt("numLamps truncated");
+     double lumensPerLamp = getDouble("lumensPerLamp truncated");
+     double candelaMultiplier = getDouble("candelaMultiplier truncated");
+     int vSize = getInt("vSize truncated");
+     int hSize = getInt("hSize truncated");
+
+     int type_ = getInt("type truncated");
+     if (error)
+         return false;
+     if (type_ <= 0 || type_ > 3) {
+         errorMsg = "unrecognized type";
+         return false;
+     }
+     CIESProfile::PhotometricType type =
+         static_cast<CIESProfile::PhotometricType>(type_);
+     assert(type == CIESProfile::PhotometricType::TYPE_C &&
+         "Only type C is supported for now");
```

Comment:
log instead of assertingt, asserts only work in Debug and crash the app, return false instead